### ISSUE TITLE
Memory leak when using ClusterClient

### ISSUE THREAD

**Issue by elqueffo** [dotnet/orleans#3299]
I noticed our memory usage climbing to 6gb on the client side when calling into our very simple grains. I managed to get a dotMemory snapshot and saw that most of it is in ProxiedMessageCenter.

I set up a small test case where i create a client like so:

            var serviceName = new Uri("fabric:/StatelessCalculatorApp/StatelessCalculatorService");
            IClusterClient client = null;
            for (int i = 0; i < 100; i++)
            {
                client = new ClientBuilder()
                    .UseConfiguration(new ClientConfiguration())
                    .AddServiceFabric(serviceName)
                    .ConfigureServices(
                        services =>
                        {
                            // Some deployments require a custom FabricClient, eg so that cluster endpoints and certificates can be configured.
                            // A pre-configured FabricClient can be injected.
                            var fabricClient = new FabricClient();
                            services.AddSingleton(fabricClient);
                        })
                    .Build();
                await client.Connect();
                //await client.Close();
                client.Dispose();
            }
            Console.ReadKey();

Not calling client.Close() (note it is commented out above), will leak exactly 100 OutsideRuntimeClient, which retains ProxiedMessageCenter. ProxiedMessageCenter contains PendingInboundMessages, which in our case was the culprit of the enormous memory usage.

It seems to me that calling Dispose should clean up after itself, and not require a call to Close().

Changing the above code to call client.Close() will still cause there to be two OutsideRuntimeClient instances, even after forcing GC. As i understand it, there should only be one?

**Comment by elqueffo** [comment]
So yesterday i added a Close() to our client code, but overnight i am again back at 2+gb of memory usage on the client.
![orleans-mem1](https://user-images.githubusercontent.com/8071240/29305661-270cce62-819a-11e7-8ad0-5e347ac94953.jpg)
Looks like ProxiedMessageCenter is using most of it, in PendingInboundMessages. Am I missing something obvious that i should be doing to stop this from growing seemingly unbounded?


![orleans-mem2](https://user-images.githubusercontent.com/8071240/29305743-873f0156-819a-11e7-81ca-a9881468f2d2.jpg)

**Comment by ReubenBond** [comment]
@elqueffo how many clients are you creating in your application and what is their lifecycle? Clients are intended to live for the duration of your application and be shared between threads.

**Comment by elqueffo** [comment]
@ReubenBond i create one client, but if it IsInitialized becomes false at any point after the initial setup, i recreate it.

In the screenshots above, there is only one active instance according to dotMemory. It's almost like the message pump thread died and messages just keep piling up in PendingInboundMessages. I haven't verified that theory though, since that task isn't monitored or available for checking through the public APIs.

**Comment by ReubenBond** [comment]
I should point out: definitely looking into this either way to fix any issues. There's one issue already fixed which will be included in 1.5.1, but I'll look deeper.

**Comment by elqueffo** [comment]
I thought I should also note that it is important for our use case that these clients don't leave anything behind when shutting down, and also that we can cleanly run multiple per process. This is because of the shared process model in service fabric, where multiple services and/or replicas run in the same host process.

**Comment by sergeybykov** [comment]
From https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/dispose-pattern:
>✓ CONSIDER providing method Close(), in addition to the Dispose(), if close is standard terminology in the area. 
When doing so, it is important that you make the Close implementation identical to Dispose and consider implementing the IDisposable.Dispose method explicitly.
```csharp
public class Stream : IDisposable {  
    IDisposable.Dispose(){  
        Close();  
    }  
    public void Close(){  
        Dispose(true);  
        GC.SuppressFinalize(this);  
    }  
}
```

**Comment by sergeybykov** [comment]
Resolved via #3306.
